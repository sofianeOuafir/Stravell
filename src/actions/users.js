import { editAuthUserPhotoURL } from "./auth";
import { editPostsUserPhotoURL } from "./posts";
import database from "./../firebase/firebase";

export const startUpdateUserProfilePicture = ({
  uid,
  userPhotoURL
}) => async dispatch => {
  let updateObject = {};

  // users/uid
  updateObject[`users/${uid}/userPhotoURL`] = userPhotoURL;

  // user-posts
  await database.ref(`user-posts/${uid}/`).once("value", function(snapshot) {
    let postKeys;
    try {
      postKeys = Object.keys(snapshot.val());
    } catch (error) {
      postKeys = [];
    }

    postKeys.forEach(key => {
      updateObject[`user-posts/${uid}/${key}/userPhotoURL`] = userPhotoURL;
    });
  });

  // posts
  await database
    .ref(`posts/`)
    .orderByChild("uid")
    .equalTo(uid)
    .once("value", function(snapshot) {
      let postKeys;
      try {
        postKeys = Object.keys(snapshot.val());
      } catch (error) {
        postKeys = [];
      }
      postKeys.forEach(key => {
        updateObject[`posts/${key}/userPhotoURL`] = userPhotoURL;
      });
    });

  // country-posts
  await database.ref(`country-posts/`).once("value", function(snapshot) {
    snapshot.forEach(function(child) {
      const countryCode = child.key;
      child.ref
        .orderByChild("uid")
        .equalTo(uid)
        .once("value", function(snapshot) {
          snapshot.forEach(function(child) {
            const postId = child.key;
            updateObject[
              `country-posts/${countryCode}/${postId}/userPhotoURL`
            ] = userPhotoURL;
          });
        });
    });
  });

  // update place-posts
  await database.ref(`place-posts/`).once("value", function(snapshot) {
    snapshot.forEach(function(child) {
      const placeId = `${child.key}`;
      child.ref
        .orderByChild("uid")
        .equalTo(uid)
        .once("value", function(snapshot) {
          snapshot.forEach(function(child) {
            const postId = child.key;
            updateObject[
              `place-posts/${placeId}/${postId}/userPhotoURL`
            ] = userPhotoURL;
          });
        });
    });
  });

  // // update region-posts
  await database.ref(`region-posts/`).once("value", function(snapshot) {
    snapshot.forEach(function(child) {
      const regionCode = child.key;
      child.ref
        .orderByChild("uid")
        .equalTo(uid)
        .once("value", function(snapshot) {
          snapshot.forEach(function(child) {
            const postId = child.key;
            updateObject[
              `region-posts/${regionCode}/${postId}/userPhotoURL`
            ] = userPhotoURL;
          });
        });
    });
  });

  return database
    .ref()
    .update(updateObject)
    .then(() => {
      dispatch(editAuthUserPhotoURL(`${userPhotoURL}?${new Date().getTime()}`));
      dispatch(
        editPostsUserPhotoURL({
          userPhotoURL: `${userPhotoURL}?${new Date().getTime()}`,
          uid
        })
      );
    })
    .catch(e => {
      console.log(e);
    });
};
