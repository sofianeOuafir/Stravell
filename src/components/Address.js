import React from "react";
import { MdMyLocation } from "react-icons/md";
import Link from "next/link";
import { slugify } from "underscore.string";

import { getPlaceIdFromLatLng } from "./../lib/utils/place";

const Address = ({ iconClassName, addressClassName, address, lat, lng }) => {
  const placeId = getPlaceIdFromLatLng({ lat, lng });
  return (
    <div className="flex align-items--center">
      <Link
        href={`/place?id=${placeId}`}
        as={`/place/${slugify(address)}/${placeId}`}
      >
        <div>
          <MdMyLocation className={iconClassName ? iconClassName : ""} />
          <span className={addressClassName ? addressClassName : ""}>
            {address}
          </span>
        </div>
      </Link>
    </div>
  );
};

export default Address;
