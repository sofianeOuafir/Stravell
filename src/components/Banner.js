import React from "react";
import { IoIosSearch, IoIosAirplane } from "react-icons/io";
import Typed from "react-typed";
import { Link as ScrollLink } from "react-scroll";

import SocialShareButtons from "./SocialShareButtons";
import Image from './Image';
class Banner extends React.Component {
  render() {
    const { imageUrl, place, placeSuggestions } = this.props;
    return (
      <div className="banner__image-container">
        <div className="content-container">
          <div className="banner__text-container" />
          <div className="banner__text">
            <h1 className="banner__website-name">
              Stravell {place ? `| ${place}` : ""}
            </h1>
            <span className="banner__description">
              A search engine for travel articles. <br /> It makes it easy for you
              to find travel articles about thousands of places
              <br />
              around the world written by some of the best travel bloggers
              worldwide.
            </span>

          </div>
          <div className="banner__typing-container" />
          <div className="banner__typing flex align-items--center">
            <IoIosSearch className="mr1" />
            <Typed
              strings={placeSuggestions}
              typeSpeed={100}
              backSpeed={100}
              loop
            />
          </div>
          <div className="banner__get_started-container">
            <ScrollLink
              href="#"
              className="button button--large button--warm-peach"
              to="searchBarInput"
              smooth="true"
              duration={600}
              offset={-100}
            >
              <div className="flex align-items--center">
                <IoIosAirplane className="mr1" /> <span>Get Started</span>
              </div>
            </ScrollLink>
          </div>

          <div className="banner__social_buttons">
            <SocialShareButtons />
          </div>
        </div>
        <Image className="fullwidth" src={imageUrl} alt="" loadingImageWidth={1026} loadingImageHeight={405} />
      </div>
    );
  }
}

export default Banner;
