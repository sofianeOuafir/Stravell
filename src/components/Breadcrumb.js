import React, { Fragment } from "react";
import Link from "next/link";

const Breadcrumb = ({ links }) => (
  <div>
    {links &&
      // linkProps would be "href" and "as" props
      links.map(({ text, ...linkProps }, index) => {
        if (linkProps.href && text) {
          return (
            <Link key={index} {...linkProps}>
              <a className="text-dark-grey no-text-decoration">
                <span className="mr1">{text}</span>
                {links.length !== index + 1 && <span className="mr1">/</span>}
              </a>
            </Link>
          );
        }
      })}
  </div>
);

export default Breadcrumb;
