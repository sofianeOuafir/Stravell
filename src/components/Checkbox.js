import React from "react";
import uuid from 'uuid';
class Checkbox extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    const id = uuid();
    return (
      <div>
        <input
          id={id}
          name={id}
          className="mr1"
          onChange={this.props.handleChange}
          type="checkbox"
          checked={this.props.checked}
        />
        <label htmlFor={id}>{ this.props.label }</label>
      </div>
    );
  }
}

export default Checkbox;
