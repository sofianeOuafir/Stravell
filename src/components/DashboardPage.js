import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import FilterablePostList from "./FilterablePostList";
import PageHeader from "./PageHeader";

export const DashboardPage = props => {
  return (
    <div>
      <PageHeader title="Dashboard" />
      <div className="content-container">
        <Link to="posts/create" className="button button--with-bottom-margin">
          Create Post
        </Link>
        <FilterablePostList
          editable={true}
          SearchBarAutoFocus={true}
          posts={props.posts}
          noPostText={`The world is looking forward to hear about your stories! :)`}
        />
      </div>
    </div>
  );
};

const mapStateToProps = ({ posts, auth }) => ({
  posts: posts.filter(post => post.uid === auth.uid),
  userName: auth.userName
});

export default connect(mapStateToProps)(DashboardPage);
