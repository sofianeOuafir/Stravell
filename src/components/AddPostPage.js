import React from "react";
import PostForm from "./PostForm";
import { connect } from "react-redux";
import { startAddPost } from "../actions/posts";
import PageHeader from './PageHeader';

class AddPostPage extends React.Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <PageHeader title="Create Post" />
        <div className="content-container">
          <PostForm onSubmit={this.onSubmit} /> 
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => {
    dispatch(startAddPost(post));
  }
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddPostPage);
