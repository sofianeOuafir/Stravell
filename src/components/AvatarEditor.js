import React, { Fragment } from "react";
import Editor from "react-avatar-editor";
import { connect } from "react-redux";
import uuid from "uuid";
import Avatar from "react-avatar";
import { IoMdCamera } from "react-icons/io";
import Modal from "react-modal";

import { uploadFile } from "./../aws/s3";
import { updateUserProfilePicture } from "./../queries/user";
import { updateProfilePicture } from "./../actions/auth";
import Loading from "./Loading";

class AvatarEditor extends React.Component {
  constructor(props) {
    super(props);
    const { userPhotoURL: currentImageUrl, uid } = props.user;
    this.state = {
      currentImageUrl,
      uploadedImage: null,
      uid,
      uploading: false,
      modalIsOpen: false
    };
  }

  closeModal = () => {
    this.setState(() => ({ modalIsOpen: false }));
  };

  onImageSave = () => {
    const { uid } = this.state;
    const { updateProfilePicture } = this.props;
    const canvas = this.editor.getImage();
    this.setState(() => ({ uploading: true, modalIsOpen: false }));

    canvas.toBlob(blob => {
      const file = new File([blob], "profile_picture", {
        type: blob.type
      });
      uploadFile({ file, location: `users/${uid}` })
        .then(data => {
          const { Location: newUserPhotoURL } = data;
          this.setState(() => ({
            uploading: false,
            currentImageUrl: `${newUserPhotoURL}?${new Date().getTime()}`
          }));

          updateUserProfilePicture({ uid, newUserPhotoURL });
          updateProfilePicture(`${newUserPhotoURL}?${new Date().getTime()}`);
        })
        .catch(e => {
          this.setState(() => ({ uploading: false }));
          alert(e);
        });
    });
  };

  onImageChange = e => {
    const uploadedImage = e.target.files[0];
    if (!uploadedImage) {
      return;
    }
    this.setState(() => ({ uploadedImage, modalIsOpen: true }));
  };

  onCancel = () => {
    this.setState(() => ({
      modalIsOpen: false
    }));
  };

  triggerFileDialog = () => {
    document.getElementById("imageInput").click();
  };

  setEditorRef = editor => (this.editor = editor);

  render() {
    const {
      uploadedImage,
      currentImageUrl,
      uploading,
      modalIsOpen
    } = this.state;
    return (
      <div
        className="relative flex align-items--center justify-content--center"
        style={{ height: 100, width: 100 }}
      >
        <Fragment>
          <Avatar
            className="absolute pointer"
            src={currentImageUrl}
            round={true}
            size={100}
            onClick={this.triggerFileDialog}
          />
          {uploading ? (
            <div className="absolute">
              <Loading size={40} />
            </div>
          ) : (
            <Fragment>
              <input
                id="imageInput"
                className="hide"
                type="file"
                onChange={this.onImageChange}
              />

              <IoMdCamera
                className="absolute pointer h2 text-white"
                onClick={this.triggerFileDialog}
              />
            </Fragment>
          )}
        </Fragment>

        <Modal
          isOpen={modalIsOpen}
          onRequestClose={this.closeModal}
          className="modal"
          contentLabel="Example Modal"
        >
          <p className="modal__title h3 favourite-font-weight">Edit Media</p>
          <Editor ref={this.setEditorRef} image={uploadedImage} />
          <div className="flex justify-content--between">
            <button className="button" onClick={this.onImageSave}>Save</button>
            <button className="button button--warm-peach" onClick={this.onCancel}>Cancel</button>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  updateProfilePicture: userPhotoURL =>
    dispatch(updateProfilePicture(userPhotoURL))
});

export default connect(
  null,
  mapDispatchToProps
)(AvatarEditor);
