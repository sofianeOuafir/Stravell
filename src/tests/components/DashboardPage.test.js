import React from 'react';
import { shallow } from 'enzyme';
import { DashboardPage } from '../../components/DashboardPage';
import users from './../fixtures/users';

test('should render DashboardPage correctly', () => {
  const user = users[0];
  user.id = user.uid
  const wrapper = shallow(<DashboardPage user={user} />);
  expect(wrapper).toMatchSnapshot();
});
