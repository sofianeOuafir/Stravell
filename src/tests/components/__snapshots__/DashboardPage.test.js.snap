// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DashboardPage correctly 1`] = `
<div>
  <PageHeader
    title="Dashboard"
  />
  <div
    className="content-container"
  >
    <Link
      className="button button--with-bottom-margin"
      replace={false}
      to="posts/create"
    >
      Create Post
    </Link>
    <Connect(FilterablePostList)
      SearchBarAutoFocus={true}
      editable={true}
      noPostText="The world is looking forward to hear about your stories! :)"
    />
  </div>
</div>
`;
